src/support/modify.c

```diff
@@ -347,86 +353,56 @@ __modify_apply(WT_SESSION_IMPL *session,
-	 * session pointer - they don't expect statistics to be incremented.
-	 */
-	if (session != NULL) {
-		for (modified = 0, i = 0; i < nentries; ++i)
-			modified += entries[i].size;
-		WT_STAT_CONN_INCR(session, cursor_modify);
-		WT_STAT_DATA_INCR(session, cursor_modify);
-		WT_STAT_CONN_INCRV(session,
-		    cursor_modify_bytes, cursor->value.size);
-		WT_STAT_DATA_INCRV(session,
-		    cursor_modify_bytes, cursor->value.size);
-		WT_STAT_CONN_INCRV(session,
-		    cursor_modify_bytes_touch, modified);
-		WT_STAT_DATA_INCRV(session,
-		    cursor_modify_bytes_touch, modified);
-	}
-
-	return (0);
+err:	__wt_scr_free((WT_SESSION_IMPL *)cursor->session, &modify);
+	return (ret);
 }
 
 /*
  * __wt_modify_apply --
  *	Apply a single set of WT_MODIFY changes to a buffer.
  */
 int
-__wt_modify_apply(
-    WT_SESSION_IMPL *session, WT_CURSOR *cursor, const void *modify)
+__wt_modify_apply(WT_CURSOR *cursor, const void *modify)
 {
-	WT_DECL_RET;
-	WT_MODIFY *entries, _entries[30];
-	size_t nentries;
+	WT_ITEM *value;
+	WT_MODIFY mod;
+	WT_SESSION_IMPL *session;
+	size_t datasz, destsz, tmp;
 	const size_t *p;
-	const uint8_t *data;
-	u_int i;
+	int nentries;
+	bool fast_path, sformat;
 
-	entries = NULL;
+	session = (WT_SESSION_IMPL *)cursor->session;
+	sformat = cursor->value_format[0] == 'S';
+	value = &cursor->value;
 
 	/*
-	 * Get the number of modify entries and get an array to hold them. The
-	 * 30 structures allocated on the stack was chosen because it's under
-	 * 2KB on the stack at 50B per WT_MODIFY structure, and 30 modifications
-	 * is a lot of changes for a single buffer.
+	 * Get the number of modify entries and set a second pointer to
+	 * reference the replacement data.
 	 */
 	p = modify;
-	memcpy(&nentries, p++, sizeof(size_t));
-	if (nentries <= WT_ELEMENTS(_entries))
-		entries = _entries;
-	else
-		WT_RET(__wt_malloc(
-		    session, nentries * sizeof(WT_MODIFY), &entries));
+	memcpy(&tmp, p++, sizeof(size_t));
+	nentries = (int)tmp;
 
-	/*
-	 * Set a second pointer to reference the change data. The modify string
-	 * isn't necessarily aligned for size_t access, copy memory to be sure.
-	 */
-	data = (uint8_t *)modify +
-	    sizeof(size_t) + (nentries * 3 * sizeof(size_t));
-	for (i = 0; i < nentries; ++i) {
-		memcpy(&entries[i].data.size, p++, sizeof(size_t));
-		memcpy(&entries[i].offset, p++, sizeof(size_t));
-		memcpy(&entries[i].size, p++, sizeof(size_t));
-		entries[i].data.data = data;
-		data += entries[i].data.size;
-	}
+	fast_path = __modify_check_fast_path(
+	    value, p, nentries, &datasz, &destsz);
 
-	ret = __modify_apply(session, cursor, entries, (int)nentries);
+	if (fast_path && !sformat)
+		return (__modify_apply_no_overlap(
+		    session, value, p, nentries, datasz, destsz));
 
-	if (entries != _entries)
-		__wt_free(session, entries);
+	WT_MODIFY_FOREACH_BEGIN(mod, p, nentries) {
+		WT_RET(__modify_apply_one(session, value, &mod, sformat));
+	} WT_MODIFY_FOREACH_END;
```

Comment:
I don't understand why we can't fast-path `sformat` values.

If we can't, is there any point in calling `__modify_check_fast_path()`?

Reply:
Fixed, by moving that logic up to the calling function.